modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+                        "for type [" + irUnaryMathNode.getExpressionCanonicalTypeName() + "]");
+            }
+        }
+    }
+
+    @Override
+    public void visitBinaryMath(BinaryMathNode irBinaryMathNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irBinaryMathNode.getLocation());
+
+        Operation operation = irBinaryMathNode.getOperation();
+        ExpressionNode irLeftNode = irBinaryMathNode.getLeftNode();
+        ExpressionNode irRightNode = irBinaryMathNode.getRightNode();
+
+        if (operation == Operation.FIND || operation == Operation.MATCH) {
+            visit(irRightNode, writeScope);
+            methodWriter.push(irBinaryMathNode.getRegexLimit());
+            visit(irLeftNode, writeScope);
+            methodWriter.invokeStatic(org.objectweb.asm.Type.getType(Augmentation.class), WriterConstants.PATTERN_MATCHER);
+
+            if (operation == Operation.FIND) {
+                methodWriter.invokeVirtual(org.objectweb.asm.Type.getType(Matcher.class), WriterConstants.MATCHER_FIND);
+            } else if (operation == Operation.MATCH) {
+                methodWriter.invokeVirtual(org.objectweb.asm.Type.getType(Matcher.class), WriterConstants.MATCHER_MATCHES);
+            } else {
+                throw new IllegalStateException("unexpected binary math operation [" + operation + "] " +
+                        "for type [" + irBinaryMathNode.getExpressionCanonicalTypeName() + "]");
+            }
+        } else {
+            visit(irLeftNode, writeScope);
+            visit(irRightNode, writeScope);
+
+            if (irBinaryMathNode.getBinaryType() == def.class ||
+                    (irBinaryMathNode.getShiftType() != null && irBinaryMathNode.getShiftType() == def.class)) {
+                methodWriter.writeDynamicBinaryInstruction(irBinaryMathNode.getLocation(), irBinaryMathNode.getExpressionType(),
+                        irLeftNode.getExpressionType(), irRightNode.getExpressionType(), operation, irBinaryMathNode.getFlags());
+            } else {
+                methodWriter.writeBinaryInstruction(irBinaryMathNode.getLocation(), irBinaryMathNode.getExpressionType(), operation);
+            }
+        }
+    }
+
+    @Override
+    public void visitStringConcatenation(StringConcatenationNode irStringConcatenationNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irStringConcatenationNode.getLocation());
+        methodWriter.writeNewStrings();
+
+        for (ExpressionNode argumentNode : irStringConcatenationNode.getArgumentNodes()) {
+            visit(argumentNode, writeScope);
+            methodWriter.writeAppendStrings(argumentNode.getExpressionType());
+        }
+
+        methodWriter.writeToStrings();
+    }
+
+    @Override
+    public void visitBoolean(BooleanNode irBooleanNode, WriteScope writeScope) {
```

Comment:
Consider calling this visitCondition.

Reply:
I'm concerned about mixing that up with conditional nodes. I agree this needs renaming, but I need to think about it so probably a follow up PR for this.